src/gwt/src/org/rstudio/studio/client/workbench/views/output/find/model/FindResult.java

```diff
@@ -58,51 +101,206 @@ public native final String getLineValue() /*-{
+               {
+                  if (replaceOnVal >= 0)
+                     difference = offVal - onVal;
+                  else
+                     difference = -1; // shouldn't get here
+                  replaceParts.add(new Pair<Boolean, Integer>(true, (replaceOn.remove(0) + difference)));
+               }
+               else
+                  replaceParts.add(new Pair<Boolean, Integer>(false, (replaceOff.remove(0))));
             }
-            else if (openTags > 0)
+         }
+   
+         String line = getLineValue();
+         // Use a counter to ensure tags are balanced.
+         int openStrongTags = 0;
+         int openEmTags = 0;
+   
+         for (int i = 0; i < line.length(); i++)
+         {
+            while (parts.size() > 0 && parts.get(0).second == i)
+            {
+               if (parts.remove(0).first)
+               {
+                  out.appendHtmlConstant("<strong>");
+                  openStrongTags++;
+               }
+               else if (openStrongTags > 0)
+               {
+                  out.appendHtmlConstant("</strong>");
+                  openStrongTags--;
+                  String replace = getReplaceValue();
+                  if (!StringUtil.isNullOrEmpty(replace))
+                  {
+                     out.appendHtmlConstant("<em>");
+                     for (int j = 0; j < replace.length(); j++)
+                        out.append(replace.charAt(j));
+                     out.appendHtmlConstant("</em>");
+                  }
+               }
+            }
+            while (replaceParts.size() > 0 && replaceParts.get(0).second == i)
+            {
+               if (replaceParts.remove(0).first)
+               {
+                  out.appendHtmlConstant("<em>");
+                  openEmTags++;
+               }
+               else if (openEmTags > 0)
+               {
+                  out.appendHtmlConstant("</em>");
+                  openEmTags--;
+               }
+            }
+            if (getRegexPreviewIndicator() || openEmTags == 0)
+               out.append(line.charAt(i));
+         }
+   
+         while (openStrongTags > 0)
+         {
+            openStrongTags--;
+            out.appendHtmlConstant("</strong>");
+            String replace = getReplaceValue();
+            if (!StringUtil.isNullOrEmpty(replace))
             {
-               out.appendHtmlConstant("</strong>");
-               openTags--;
+               out.appendHtmlConstant("<em>");
+               for (int j = 0; j < replace.length(); j++)
+                  out.append(replace.charAt(j));
+               out.appendHtmlConstant("</em>");
             }
          }
-         out.append(line.charAt(i));
+   
+         while (openEmTags > 0)
+         {
+            openEmTags--;
+            out.appendHtmlConstant("</em>");
+         }
       }
 
-      while (openTags > 0)
+      return out.toSafeHtml();
+   }
+
+   public final SafeHtml getLineReplaceHTML()
+   {
```

Comment:
Is there any part of this function and the previous one that could be extracted into a shared helper function?

Reply:
After working through the bugs discovered here, I don't think there can be a shared helper function. The two functions are doing similar things, but there are small differences that prevent the code from being shared.

Reply:
Created project card for review.